File: 1-yahtzee-java/src/main/java/yahtzee/Categories.java
package yahtzee;

public class Categories {
    private ConsoleNotifier notifier;
    private ScoresHistory scoresHistory;
    private DiceRoller diceRoller;
    private Reruns reruns;

    public Categories(
            ConsoleNotifier notifier,
            DiceRoller diceRoller,
            ScoresHistory scoresHistory,
            Reruns reruns
    ) {
        this.notifier = notifier;
        this.scoresHistory = scoresHistory;
        this.diceRoller = diceRoller;
        this.reruns = reruns;
    }

    public void play(int numReruns) {
        for (Category category : Category.values()) {
            this.notifier.notifyCurrentCategory(category);
            rollAll();
            this.reruns.doReruns(numReruns);
            int score = category.computeScore(this.diceRoller.lastRolledDice());
            this.scoresHistory.annotateScore(category, score);
            this.notifier.notifyCategoryScore(category, score);
        }
    }

    private void rollAll() {
        this.diceRoller.roll(Dice.values());
        this.notifier.notifyRolledDice(this.diceRoller.lastRolledDice());
    }
}

File: 02-yahtzee-php/src/Yahtzee/Categories.php
<?php


namespace Yahtzee;


class Categories
{
    /** @var UserInterface */
    private $userInterface;

    /** @var DiceRoller */
    private $diceRoller;

    /** @var ReRuns */
    private $reRuns;

    /** @var ScoresSummary */
    private $scoresSummary;

    /**
     * @param UserInterface $userInterface
     * @param DiceRoller $diceRoller
     * @param ReRuns $reRuns
     * @param ScoresSummary $scoresSummary
     */
    public function __construct(
        UserInterface $userInterface,
        DiceRoller $diceRoller,
        ReRuns $reRuns,
        ScoresSummary $scoresSummary
    ) {
        $this->userInterface = $userInterface;
        $this->diceRoller = $diceRoller;
        $this->reRuns = $reRuns;
        $this->scoresSummary = $scoresSummary;
    }

    /**
     * @param int $numReRuns
     */
    public function play($numReRuns)
    {
        foreach (Category::all() as $category) {
            $this->userInterface->printCategory($category);
            $this->diceRoller->rollAll();
            $this->userInterface->printDiceLine($this->diceRoller->lastRollResult());
            $this->reRuns->doReRuns($numReRuns);
            $categoryScore = $category->calculateScore($this->diceRoller->lastRollResult());
            $this->scoresSummary->registerScore($category, $categoryScore);
            $this->userInterface->printCategoryScore($category, $categoryScore);
        }
    }

    public function printScoresSummary()
    {
        $this->userInterface->printScoresSummary($this->scoresSummary);
    }
}
